test/com/amazon/ionelement/EquivTestCase.kt

```diff
@@ -83,8 +80,98 @@ data class EquivTestCase(val left: String, val right: String, val isEquiv: Boole
         checkIt(leftElement, rightElement)
         checkIt(rightElement, leftElement)
 
+        // Try using a proxy to make sure that equivalence is not tied to a particular implementation.
+        val leftProxy = leftElement.proxy()
+        val rightProxy = rightElement.proxy()
+
+        Assertions.assertEquals(leftElement, leftProxy)
+        Assertions.assertEquals(rightProxy, rightElement)
+
+        checkIt(leftElement, rightProxy)
+        checkIt(leftProxy, rightElement)
+        checkIt(leftProxy, rightProxy)
+
         // Adding metas to one side should not have any effect
         checkIt(leftElement, rightElement.withMeta("foo", 123))
         checkIt(rightElement, leftElement.withMeta("bar", 456))
     }
+
+    /**
+     * Returns an anonymous class that is a proxy for this IonElement.
+     */
+    private inline fun <reified T : IonElement> T.proxy(): T {
+        val delegate = this
+        return when (delegate) {
+            is BoolElement -> object : BoolElement by delegate {
+                override fun equals(other: Any?): Boolean = delegate == other
+                override fun hashCode(): Int = delegate.hashCode()
+                override fun toString(): String = "AnonymousProxy($delegate)"
```

Comment:
Since the override logic across each anonymous class for each `IonElement` types is consistent, should we consider abstract the common logic?

Reply:
I don't think it's particularly important here, but if you feel strongly about it, I'll do it.

Reply:
> I don't think it's particularly important here, but if you feel strongly about it, I'll do it.

Thanks for considering the feedback, I just see this could potentially enhance the maintainability. However, if there are other higher priorities at this time we can postpone this suggested change.